### PR TITLE
feat(core): middleware runs once for matching route

### PR DESCRIPTION
**Pull request by rychardvale**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently, routes that has dynamic params can trigger a middleware multiple times for a single request if matches another static route inside the controller.

Issue Number: [#1628](https://github.com/nestjs/nest/issues/1628)

## What is the new behavior?
Routes will only trigger the middleware for the exact path match.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: packages/core/middleware/builder.ts [hunk 1]
```diff
@@ -82,5 +83,25 @@ export class MiddlewareBuilder implements MiddlewareConsumer {
         .flatten()
         .toArray();
     }
+
+    private removeOverlappedRoutes(routes: RouteInfo[]) {
+      const regexMatchParams = /(:[^\/]*)/g;
+      const wildcard = '([^/]*)';
+      const routesWithRegex = routes
+        .filter(route => route.path.indexOf(':') >= 0)
```
Comment (micalevisk) [inline packages/core/middleware/builder.ts:91]:
```suggestion
        .filter(route => route.path.includes(':'))
```

File: packages/core/middleware/builder.ts [hunk 2]
```diff
@@ -82,5 +83,25 @@ export class MiddlewareBuilder implements MiddlewareConsumer {
...
+          path: route.path,
+          regex: new RegExp(
+            '^(' + route.path.replace(regexMatchParams, wildcard) + ')$',
+            'g',
+          ),
+        }));
+      return routes.filter(route => {
+        const routeMatch = routesWithRegex.find(
+          v => route.path !== v.path && route.path.match(v.regex),
+        );
```
[the remark this answers is not in the document]
Comment (rychardvale) [reply packages/core/middleware/builder.ts:102]: Absolutely! Will do.

**Review by micalevisk — APPROVED** [review]

File: packages/core/middleware/builder.ts [hunk 3]
```diff
@@ -82,5 +83,27 @@ export class MiddlewareBuilder implements MiddlewareConsumer {
...
+            'g',
+          ),
+        }));
+      return routes.filter(route => {
+        const isOverlapped = (v: { path: string; regex: RegExp }) => {
+          return route.path !== v.path && route.path.match(v.regex);
+        };
+        const routeMatch = routesWithRegex.find(isOverlapped);
+
+        if (routeMatch === undefined) return route;
```
Comment (Tony133) [inline packages/core/middleware/builder.ts:105]:
```suggestion
        if (routeMatch === undefined) {
         return route;
        }
```

**Review by Tony133 — CHANGES_REQUESTED** [review]
A little convention/syntax fix while otherwise fine, but let's see what Kamil thinks.

**Review by Tony133 — APPROVED** [review]

**Review by hebertcisco — APPROVED** [review]

**Pull request merged**